Extended block collection

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This PR extends the functionality of the `BlockCollector` class, replacing the previous PR #11614. The key new feature is that the check whether a node should be added to the current block may now depend both on the ``filter_fn`` (which specifies _which_ nodes should be collected) _and the block collected so far_. The logic of the `StarPreRouting` pass has been simplified to use the 
`BlockCollector` functionality instead of essentially implementing this in-place. In addition to avoiding code duplication, this also allows to extend the `StarPreRouting` pass with additional options such as using `DAGDependency` dag representation or collecting star-shaped blocks nodes from the back of the circuit (this is not done in this PR since it requires #12418). Note that for `StarPreRouting` the logic of whether a node can be added to the block collected so far depends on the block, i.e. whether it remains star-shaped. 

The new class `Block` defines the abstract API that a block class should satisfy in order to use it for block collection. The main method is ``append_node`` which specifies whether a node can be added to the current block. The method ``size`` returns the size of the block which is useful when filtering blocks based on the ``min_block_size`` argument. The method ``get_nodes`` is primarily needed for backwards compatibility and should return the list of nodes in the block (however it would not be called when using block collection with a specified block class). The method ``reverse`` needs to be implemented to support collecting nodes from the back of the circuit. The method ``split`` needs to be implemented to support splitting collected blocks into smaller sub-blocks.

The new class `DefaultBlock` implements the simple functionality that can be used when the decision of whether to add a node in the current block only depends on ``filter_fn`` and not on the block collected so far. 

The block collection class and functionality are backward compatible, any code that worked before should remain working. (This is why ``collect_all_matching_blocks`` returns a list of blocks where each block is either a subclass of `Block` or a list of nodes). However, the previously public methods ``collect_matching_block`` and ``collect_all_matching_blocks`` in `StarPreRouting` were completely removed (making this specific change not backward-compatible). It was an oversight to make this methods public in the first place, and unfortunately these methods don't make much sense outside of `BlockCollector`.

alexanderivrii committed:
extending the functionality of the BlockCollector class

alexanderivrii committed:
test new block collection functionality

alexanderivrii committed:
release notes

alexanderivrii committed:
pylint fixes

alexanderivrii committed:
Merge branch 'main' into block-collection

henryzou50 commented on qiskit/dagcircuit/collect_blocks.py:
```diff
@@ -25,6 +27,139 @@
+                sub-blocks).
+
+        """
+        raise NotImplementedError
```
I like how well-documented this abstract base class is and it is well designed. One potential improvement here is clarifying the return times in order to make the return types consistent. 

For methods like `get_nodes`, `reverse`, and `split`, which could return `None`, it might be more intuitive to always return a list. Here an empty list could signify the same thing as `None`, but it would also make the return type consistent, and thus allow the methods to be used in a wider range of contexts without having to check for `None`.

alexanderivrii replied:
I think treating an empty list of nodes as `None` might be misleading, and I think it's safer to return `None` in the case that a certain method for a given `Block` type is not supported. I clarified the return types in [1bab966](https://github.com/Qiskit/qiskit/pull/12498/commits/1bab966e6b35794fdcee062e54ec1ddd38066c49).

henryzou50 commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -11,34 +11,31 @@
 
 
-class StarBlock:
+class StarBlock(Block):
```
From an architectural standpoint, I think it makes more sense for `StarBlock` class to be placed in the `collect_blocks.py`. Here are some reasons for moving it:
- Separation of concerns: `collect_blocks,py is focused on defining different types of blocks and the logic for collecting them. Since `StarBlock is a specific type of block, its definition fits naturally within this module.
- Reusability: It might be useful to use `StarBlock` outside of `star_prerouting` and having `StarBlock` in `collect_blocks.py` makes it easier to reuse. For instance, If other passes or functionalities require star-shaped block logic, they can import `StarBlock` from a centralized module.

alexanderivrii replied:
This and other comments regarding `star_preroute.py` are no longer relevant after in [1a5ef91](https://github.com/Qiskit/qiskit/pull/12498/commits/1a5ef91b27d920b8c582bac2cab10805a81aec19) I've reverted the changes to this file, in order not to interfere with #12761.

I completely agree that it would be best to create a "library" of useful and reusable block types (like `DefaultBlock`). I am not completely convinced that specifically `StarBlock` is going to be useful beyond the `StarPrerouting` pass, so I am not sure we should expose this particular block type. In addition, I have locally tried to improve the quality of `StarPreroute` pass by experimenting with different variants of the `StarBlock` (for example only collecting stars with non-repeating non-center nodes, the intuition being that repeating nodes prevent mapping a star to a line, or only collecting stars consisting of pairwise commuting gates). Unfortunately, I did not find anything that made `StarPrerouting` better on say QAOA circuits, however I do find likely that a good star block for the star prerouting pass would be very specific to this pass and thus should be packaged together with the pass.

henryzou50 commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -288,10 +166,26 @@ def determine_star_blocks_processing(
+            filter_fn=filter_fn,
+            block_class=StarBlock,
+            output_nodes=False,
+            min_block_size=min_block_size,
```
Since the `split` method for the `Starblock` does not do anything at the moment, and `collect_all_matching_blocks` defaults `split_block=True`, we should add the parameter that `split_blocks=False` to make it clear that we are not splitting at this point.

henryzou50 replied:
This also makes me a bit confused on why we did not catch a `DAGCircuitError` in our testing of `star_prerouting`during the section of `collect_all_matching_block`:

```python
        if split_blocks or split_layers:
            tmp_blocks = []
            for block in matching_blocks:
                sub_blocks = block.split(split_blocks=split_blocks, split_layers=split_layers)
                if sub_blocks is None:
                    raise DAGCircuitError(
                        "The block class does not implement the method ``split``."
                    )
                tmp_blocks.extend(sub_blocks)
            matching_blocks = tmp_blocks
 ```

henryzou50 replied:
Oh I see, we do not catch the `DAGCircuitError` here as the `split` method for `StarBlock` always returns `[self]`, which is not `None`. Still we should have `split_blocks=False` for this section to avoid confusion and unnecessary compute.

alexanderivrii replied:
Yeah, I agree, the previous handling was sloppy, as you have pointed out.

henryzou50 commented on qiskit/transpiler/passes/routing/star_prerouting.py:
```diff
@@ -70,10 +67,31 @@ def append_node(self, node):
+        do anything and simply returns the list consisting of this
+        very block.
+        """
+        return [self]
```
We might want to include in the release notes that the `split` method for `StarBlock` acts as a placeholder and is not fully implemented yet.

alexanderivrii replied:
I agree, though this is no longer relevant at the moment.

henryzou50 reviewed (commented):
Overall, the API changes LGTM, and is well designed. I left a few comments about the structure and details about the `split` method.

alexanderivrii committed:
Merge branch 'main' into block-collection

alexanderivrii committed:
Undoing changes to StarPreroute transpiler pass

It is better to merge 12761 first, and then think how to integrate the
extended block collection into the new Rust-based flow.

alexanderivrii commented:
Sorry for the late reply. I think at this point it is better to merge #12761 first, and then to think how to best make use of the extended block collection class within `StarPreroute`. So for now I have reverted the changes to the `StarPreroute` pass (and updated the release notes accordingly), see [1a5ef91](https://github.com/Qiskit/qiskit/pull/12498/commits/1a5ef91b27d920b8c582bac2cab10805a81aec19).

alexanderivrii committed:
adding explicit type hints for methods in Block API

alexanderivrii commented:
Hi @henryzou50, can you please take another look at this PR and see what's missing? Thanks!

alexanderivrii committed:
Merge branch 'main' into block-collection

alexanderivrii commented:
I am closing this PR since it's stale and should be reimplemented in rust.
